src/simples3/credential.rs

```diff
@@ -309,180 +314,178 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
 }
 
 /// Provides AWS credentials from a resource's IAM role.
-pub struct IamProvider;
+pub struct IamProvider {
+    client: Client<HttpConnector>,
+    handle: Handle,
+}
 
-impl ProvideAwsCredentials for IamProvider {
-    fn credentials(&self) -> Result<AwsCredentials, CredentialsError> {
-        let mut client = Client::new();
-        //XXX: this is crappy, but this blocks on non-EC2 machines like
-        // our mac builders.
-        client.set_read_timeout(Some(StdDuration::from_secs(2)));
-        var("AWS_IAM_CREDENTIALS_URL")
-            .or_else(|_| {
-                // First get the IAM role
-                let mut address = "http://169.254.169.254/latest/meta-data/iam/security-credentials".to_string();
-                let mut response;
-                match client.get(&address)
-                    .header(Connection::close()).send() {
-                        Err(_) => return Err(CredentialsError::new("Couldn't connect to metadata service")), // add Why?
-                        Ok(received_response) => response = received_response
-                    };
-
-                let mut body = String::new();
-                if let Err(_) = response.read_to_string(&mut body) {
-		    return Err(CredentialsError::new("Didn't get a parsable response body from metadata service"));
+impl IamProvider {
+    pub fn new(handle: &Handle) -> IamProvider {
+        IamProvider {
+            client: Client::new(handle),
+            handle: handle.clone(),
+        }
+    }
+
+    fn iam_role(&self) -> Box<Future<Item=String, Error=CredentialsError>> {
+        // First get the IAM role
+        let address = "http://169.254.169.254/latest/meta-data/iam/security-credentials";
+        let mut req = Request::new(Method::Get, address.parse().unwrap());
+        req.headers_mut().set(Connection::close());
+        let response = self.client.request(req).and_then(|response| {
+            response.body().fold(Vec::new(), |mut body, chunk| {
```

Comment:
Does hyper not have a helper method for "collect the response body into a Vec of bytes"?